src/include/migraphx/op/unpack_int4.hpp

```diff
@@ -62,30 +62,53 @@ struct unpack_int4
     {
         check_shapes{inputs, *this}.same_dims().has(1);
         auto in_shape = inputs.front();
-        if(in_shape.type() != migraphx::shape::uint8_type)
+        if(in_shape.type() != migraphx::shape::int8_type and
+           in_shape.type() != migraphx::shape::uint8_type)
         {
-            MIGRAPHX_THROW("UNPACK_INT4: Only Unsigned Int8 type is supported for unpacking");
+            MIGRAPHX_THROW("UNPACK_INT4: Only Int8 or Uint8 is supported for unpacking");
         }
         auto new_lens = in_shape.lens();
         new_lens[axis] *= 2;
-        return {migraphx::shape::uint8_type, new_lens};
+        return {in_shape.type(), new_lens};
     }
 
     argument compute(const shape& output_shape, std::vector<argument> args) const
     {
+        auto input    = args.front();
+        auto in_shape = input.get_shape();
+
         argument result{output_shape};
-        auto in_shape = args.front().get_shape();
-        auto input    = args.at(0).get<uint8_t>();
-        auto output   = result.get<uint8_t>();
-        par_for(in_shape.elements(), [&](auto i) {
-            auto data_idx           = in_shape.multi(i);
-            auto out_data_multi_idx = data_idx;
-            out_data_multi_idx[axis] *= 2;
-            auto input_val = input[data_idx];
-            // mask first 4 bits, packing is assumed to be little endian
-            output[out_data_multi_idx] = uint8_t(0x0F) & input_val;
-            out_data_multi_idx[axis] += 1;
-            output[out_data_multi_idx] = input_val >> 4; // NOLINT(hicpp-signed-bitwise)
+
+        visit_all(result, input)([&](auto out, auto inp) {
+            par_for(in_shape.elements(), [&](auto i) {
+                using type    = typename decltype(out)::value_type;
+                auto data_idx = in_shape.multi(i);
```

Comment:
Similar type check here:

```cpp
if constexpr(sizeof(type) != 1 or not std::is_integral<type>{})
    MIGRAPHX_THROW("Unsupported type");
```

Reply:
(Done slightly differently, that we don't need to check for this).